### PR TITLE
Allow to configure link_flags in env

### PR DESCRIPTION
**Pull request by hhugo**
The main motivation is build static binaries based on a `static` profile.
And do something like
` (env (static (link_flags (:standard -ccopt -static))))`

- [x] allow to confgure link_flags in env
- [x] add some tests

### CODE REVIEW

File: xdg.opam
Comment (rgrinberg) [inline xdg.opam:30]: How did this change sneak in?

**Comment by rgrinberg** [comment]
Implementation looks good to me. Some tests would be very useful.

File: xdg.opam
Comment (hhugo) [reply xdg.opam:30]: I've no idea. I'll remove it

Comment (hhugo) [reply xdg.opam]: removed. the diff reappears when I do `./dune.exe runtest --auto-promote`

Comment (rgrinberg) [reply xdg.opam]: I wonder if it's because your bootstrapped version of dune is old? If not, then we should really fix this issue.

**Review by rgrinberg — APPROVED** [review]
Looks good. Do you have merge rights or do I need to merge myself?

**Comment by rgrinberg** [comment]
PS: needs CHANGES + docs.

**Comment by hhugo** [comment]
We should also wait on the CI as there might be diffs in other tests due to this PR

**Comment by hhugo** [comment]
There was a conflict with a PR recently merged https://github.com/ocaml/dune/pull/5185. I change my implementation. @voodoos would you mind taking a look ?

**Comment by hhugo** [comment]
I'm not sure what are the CI failures

**Comment by voodoos** [comment]
I just merged another PR with small conflicting changes, could you rebase one more time before I review ?
The main change is the last argument of `Cxx_flags.get_flags` that is no longer a `Path.Build.t` but the current `Context.t`.

Also after a brief look I think you loosed two checks in the rebase:
```
match Dune_project.use_standard_c_and_cxx_flags project with
        | Some true when Buildable.has_foreign_cxx exes.buildable -> ... get_flags ...
```

This default flags are only added is the dune_project option is Some true and the current target has c++ stubs

**Comment by hhugo** [comment]
rebased and fixed

**Comment by hhugo** [comment]
Alright, the CI seems to be happy now !

File: src/dune_rules/exe_rules.ml
Comment (voodoos) [inline src/dune_rules/exe_rules.ml:176]: I feel like a better name would be:
```suggestion
      let use_standard_cxx_flags =
```

File: src/dune_rules/super_context.ml
Comment (voodoos) [inline src/dune_rules/super_context.ml:416]: I feel like this is not exactly the expected behavior: `standard_cxx_flags` should be the `:standard` set of flag when expanding the `link_flags` field from the `env`. The result of this expansion is then the `:standard` set of flags for the executable stanza `link_flags` field.

This way it is possible to override the defautl cxx flags from the `link_flags` field in the `env` stanza.

```
default_cxx_flags
  |
  └-- [`:standard` for base `env` `link_flags` field expansion]
      |
      └-- [`:standard` for nested `env` `link_flags` field expansion]
          |
          :
          |
          └-- [`:standard` for `executable` `link_flags` field expansion]
```

**Review by voodoos — CHANGES_REQUESTED** [review]

File: src/dune_rules/super_context.ml
Comment (hhugo) [reply src/dune_rules/super_context.ml:416]: The logic you added depend on properties of executables, env doesn't have access to it

File: src/dune_rules/exe_rules.ml
Comment (hhugo) [reply src/dune_rules/exe_rules.ml]: renamed

File: doc/dune-files.rst
Comment (christinerose) [inline doc/dune-files.rst:1543]:
```suggestion
- ``(link_flags <flags>)`` to specify flags to OCaml when linking an
```

**Review by christinerose — CHANGES_REQUESTED** [review]
Suggested OCaml capitalization change

File: src/dune_rules/super_context.ml
Comment (voodoos) [reply src/dune_rules/super_context.ml:416]: Yes but can't we work around that by delaying the expansion or using a placeholder ?

I think it is important to preserve the philosophy of the [ordered set language](https://dune.readthedocs.io/en/stable/concepts.html#ordered-set-language) that allows full control over the built-in flags with the `:standard` set.

Comment (hhugo) [reply src/dune_rules/super_context.ml:416]: I'd like other dune devs to take a look at this.
Delaying evaluation seems to be a lot of work,  I don't know the codebase enough to understand the implication

I'm not sure what a placeholder would look like in practice.

@rgrinberg , @jeremiedimino  can you look at this ?

**Review by christinerose — APPROVED** [review]
The doc/dune-files.rst changes look great!

File: src/dune_rules/super_context.ml
Comment (jeremiedimino) [reply src/dune_rules/super_context.ml:416]: @voodoos has been thinking about the handling of flags a lot though some of his recent work, so I trust his judgement here.

Everything is versioned in Dune, so changing things after the fact is more difficult. For this reason we should aim to do the right thing from the start, even if it is more work.

Comment (hhugo) [reply src/dune_rules/super_context.ml:416]: @jeremiedimino, yet I don't how to implement such behavior. My hunch is that the placeholder doesn't work in practice and I'm not sure what are the implications of delaying the expansion of `: standard` in flags.

Comment (hhugo) [reply src/dune_rules/super_context.ml]: I've tried something, that is duplicate link_flags and use the existing inheritance. @voodoos how does that look to you ?

Comment (hhugo) [reply src/dune_rules/super_context.ml]: gentle ping

Comment (voodoos) [reply src/dune_rules/super_context.ml]: It looks like a reasonable solution. There is just one thing we should check: getting the default cxx flags requires that Dune run the C-compiler and preprocess a header file to create the `_build/default/.dune/ccomp/ccomp` file. It should be good but I think we should check (via a test) that this file is not created if no cxx stubs are built.

Also we should have a test reflecting the cascading behaviour. It is a bit painful to do because the C++ flags will differ depending on the context; what I did in my tests was grepping with a `or` : 

```
  $ GCC_LF_LIB=" -cclib -lstdc++ -cclib -shared-libgcc"
  $ Clang_LF_LIB=" -cclib -lc++"
  $ Msvc_LF_LIB=""

  $ dune rules main.exe  | tr -s '\n' ' ' |
  > grep -ce "Main.cmx$GCC_LF_LIB)\|Main.cmx$Clang_LF_LIB)\|Main.cmx$Msvc_LF_LIB)"
  1
```

It is a bit fragile but works reasonably good.

Comment (voodoos) [reply src/dune_rules/super_context.ml]: Also it might be useful for future readers to document the reason of the split in `link_flags.ml`.

Comment (hhugo) [reply src/dune_rules/super_context.ml]: I've added more test, and fix an existing test at the same time.

File: doc/dune-files.rst
Comment (christinerose) [reply doc/dune-files.rst:1543]: Could you please update this capitalization for OCaml before you merge? We're trying to standardize formatting like this across documentation.

**Review by christinerose — CHANGES_REQUESTED** [review]

**Review by voodoos — APPROVED** [review]
Looks good to me. Thanks for adding the tests !
Can you rebase and sign your commits ?

**Comment by hhugo** [comment]
I've rebased, rearranged and signed my commits

**Comment by voodoos** [comment]
> I've rebased, rearranged and signed my commits

It is still out-of-date with `main` apparently.

**Pull request merged**
